### PR TITLE
*: Fix the slow log format for parsing correctly

### LINKED ISSUE #27895: parse slow log return wrong result
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```sql
# Time: 2021-09-08T14:39:54.506967433+08:00
# Txn_start_ts: 427578666238083075
# User@Host: root[root] @ 172.16.0.0 [172.16.0.0]
# Conn_ID: 40507
# Query_time: 25.571605962
# Parse_time: 0.002923536
# Compile_time: 0.006800973
# Rewrite_time: 0.002100764
# Optimize_time: 0
# Wait_TS: 0.000015801
# Prewrite_time: 25.542014572 Commit_time: 0.002294647 Get_commit_ts_time: 0.000605473 Commit_backoff_time: 12.483 Backoff_types: [tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss tikvRPC regionMiss] Write_keys: 624 Write_size: 172064 Prewrite_region: 60
# DB: rtdb
# Is_internal: false
# Digest: 124acb3a0bec903176baca5f9da00b4e7512a41c93b417923f26502edeb324cc
# Num_cop_tasks: 0
# Mem_max: 856544
# Prepared: false
# Plan_from_cache: false
# Plan_from_binding: false
# Has_more_results: false
# KV_total: 86.635049185
# PD_total: 0.015486658
# Backoff_total: 100.054
# Write_sql_response_total: 0
# Succ: true
"INSERT INTO ...;
```

### 2. What did you expect to see? (Required)

```sql
>select Write_keys from `SLOW_QUERY` where time>= '2021-09-08 14:38:00' and time <= '2021-09-08 14:40:50' and txn_start_ts=427578666238083075;
+------------+
| Write_keys |
+------------+
| 624        |
+------------+
```

### 3. What did you see instead (Required)

```sql
>select Write_keys from `SLOW_QUERY` where time>= '2021-09-08 14:38:00' and time <= '2021-09-08 14:40:50' and txn_start_ts=427578666238083075;
+------------+
| Write_keys |
+------------+
| 0          |
+------------+
```

### 4. What is your TiDB version? (Required)

```sql
+-------------------------------------------------------------------+
| tidb_version()                                                    |
+-------------------------------------------------------------------+
| Release Version: v5.1.1                                           |
| Edition: Community                                                |
| Git Commit Hash: 797bddd25310ed42f0791c8eccb78be8cce2f502         |
| Git Branch: heads/refs/tags/v5.1.1                                |
| UTC Build Time: 2021-07-28 11:00:26                               |
| GoVersion: go1.16.4                                               |
| Race Enabled: false                                               |
| TiKV Min Version: v3.0.0-60965b006877ca7234adaced7890d7b029ed1306 |
| Check Table Before Drop: false                                    |
+-------------------------------------------------------------------+
```

### PR DESCRIPTION
**Pull request by sylzd**
### What problem does this PR solve?

Issue Number: close #27895 

Problem Summary:

### What is changed and how it works?

Proposal: [xxx](url) 

What's Changed:

How it Works:

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

**Comment by bb7133** [comment]
PTAL the test case of `TestString`(`util/execdetails/execdetails_test.go`):

```
497893 [2021-09-11T07:55:48.654Z]                          -Cop_time: 1.003 Process_time: 2.005 Wait_time: 1 Backoff_time: 1 Request_count: 1 Prewrite_time: 1 Commit_time: 1 Get_commit_ts_time: 1 Commit_backoff_tim       e: 1 Backoff_types: [backoff1 backoff2] Resolve_lock_time: 1 Local_latch_wait_time: 1 Write_keys: 1 Write_size: 1 Prewrite_region: 1 Txn_retry: 1 Process_keys: 10 Total_keys: 100 Rocksdb_delete_skipped_count       : 1 Rocksdb_key_skipped_count: 1 Rocksdb_block_cache_hit_count: 1 Rocksdb_block_read_count: 1 Rocksdb_block_read_byte: 100
497894 [2021-09-11T07:55:4update8.654Z]                          +Cop_time: 1.003 Process_time: 2.005 Wait_time: 1 Backoff_time: 1 Request_count: 1 Prewrite_time: 1 Commit_time: 1 Get_commit_ts_time: 1 Commit_backoff_tim       e: 1 Backoff_types: [backoff1,backoff2] Resolve_lock_time: 1 Local_latch_wait_time: 1 Write_keys: 1 Write_size: 1 Prewrite_region: 1 Txn_retry: 1 Process_keys: 10 Total_keys: 100 Rocksdb_delete_skipped_count       : 1 Rocksdb_key_skipped_count: 1 Rocksdb_block_cache_hit_count: 1 Rocksdb_block_read_count: 1 Rocksdb_block_read_byte: 100
```
The the comma is added for '[backoff1 backoff2]', is it required? If not we'd better keep the original format unchanged.

**Comment by sylzd** [comment]
> PTAL the test case of `TestString`(`util/execdetails/execdetails_test.go`):
> 
> ```
> 497893 [2021-09-11T07:55:48.654Z]                          -Cop_time: 1.003 Process_time: 2.005 Wait_time: 1 Backoff_time: 1 Request_count: 1 Prewrite_time: 1 Commit_time: 1 Get_commit_ts_time: 1 Commit_backoff_tim       e: 1 Backoff_types: [backoff1 backoff2] Resolve_lock_time: 1 Local_latch_wait_time: 1 Write_keys: 1 Write_size: 1 Prewrite_region: 1 Txn_retry: 1 Process_keys: 10 Total_keys: 100 Rocksdb_delete_skipped_count       : 1 Rocksdb_key_skipped_count: 1 Rocksdb_block_cache_hit_count: 1 Rocksdb_block_read_count: 1 Rocksdb_block_read_byte: 100
> 497894 [2021-09-11T07:55:48.654Z]                          +Cop_time: 1.003 Process_time: 2.005 Wait_time: 1 Backoff_time: 1 Request_count: 1 Prewrite_time: 1 Commit_time: 1 Get_commit_ts_time: 1 Commit_backoff_tim       e: 1 Backoff_types: [backoff1,backoff2] Resolve_lock_time: 1 Local_latch_wait_time: 1 Write_keys: 1 Write_size: 1 Prewrite_region: 1 Txn_retry: 1 Process_keys: 10 Total_keys: 100 Rocksdb_delete_skipped_count       : 1 Rocksdb_key_skipped_count: 1 Rocksdb_block_cache_hit_count: 1 Rocksdb_block_read_count: 1 Rocksdb_block_read_byte: 100
> ```
> 
> The comma is added for '[backoff1 backoff2]', is it required? If not we'd better keep the original format unchanged.

+1, I prefer comma format, otherwise, it makes the line hard to parse.  @TszKitLo40 I've seen that's your case, is there any reason that we should keep the old format?

In addtion, there other other places add comma like `tidb/store/copr/coprocessor.go`

```go
func (worker *copIteratorWorker) logTimeCopTask(costTime time.Duration, task *copTask, bo *Backoffer, resp *tikvrpc.Response) {
	logStr := fmt.Sprintf("[TIME_COP_PROCESS] resp_time:%s txnStartTS:%d region_id:%d store_addr:%s", costTime, worker.req.StartTs, task.region.GetID(), task.storeAddr)
	if bo.GetTotalSleep() > minLogBackoffTime {
		backoffTypes := strings.Replace(fmt.Sprintf("%v", bo.TiKVBackoffer().GetTypes()), " ", ",", -1)
		logStr += fmt.Sprintf(" backoff_ms:%d backoff_types:%s", bo.GetTotalSleep(), backoffTypes)
	}
```

File: util/execdetails/execdetails.go [hunk 1]
```diff
@@ -149,7 +149,7 @@ func (d ExecDetails) String() string {
 			parts = append(parts, CommitBackoffTimeStr+": "+strconv.FormatFloat(time.Duration(commitBackoffTime).Seconds(), 'f', -1, 64))
 		}
 		if len(commitDetails.Mu.BackoffTypes) > 0 {
-			parts = append(parts, BackoffTypesStr+": "+fmt.Sprintf("%v", commitDetails.Mu.BackoffTypes))
+			parts = append(parts, BackoffTypesStr+": "+strings.Replace(fmt.Sprintf("%v", commitDetails.Mu.BackoffTypes), " ", ",", -1))
```
Comment (crazycs520) [inline util/execdetails/execdetails.go:152]:
```
Backoff_types: [tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss]
```

Actually, there are only 2 kinds of backoff: `regionMiss` and `regionMiss`, there are too much duplicate information. So I want to change the output to `backoff_type*count`:

```sql
Backoff_types: [tikvRPC*10,regionMiss*10]
``` 

And if the count is 1, just ignore it, such as:

```sql
Backoff_types: [tikvRPC*10,regionMiss*10,pdRPC]
``` 

@bb7133 @breeswish what do you think?

File: infoschema/tables_test.go [hunk 2]
```diff
@@ -606,7 +606,34 @@ func prepareSlowLogfile(c *C, slowLogFileName string) {
...
+# Txn_start_ts: 427578666238083075
+# User@Host: root[root] @ 172.16.0.0 [172.16.0.0]
+# Conn_ID: 40507
+# Query_time: 25.571605962
+# Parse_time: 0.002923536
+# Compile_time: 0.006800973
+# Rewrite_time: 0.002100764
+# Optimize_time: 0
+# Wait_TS: 0.000015801
+# Prewrite_time: 25.542014572 Commit_time: 0.002294647 Get_commit_ts_time: 0.000605473 Commit_backoff_time: 12.483 Backoff_types: [tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss,tikvRPC,regionMiss] Write_keys: 624 Write_size: 172064 Prewrite_region: 60
```
Comment (crazycs520) [inline infoschema/tables_test.go:620]:
```sql
Backoff_types: [tikvRPC regionMiss]
```

For the old version of slow log, the parse will still be wrong, I think we should fix this too.

Comment (sylzd) [reply infoschema/tables_test.go:620]: It sounds reasonable, but is there really happen that the new parser working on the old version of slow log? 🤔

Comment (crazycs520) [reply infoschema/tables_test.go:620]: It will happen, after TiDB upgrades to the new version, it still can parse the old slow log. The old slow log will not be removed unless the human manual deletes it.

Comment (breeswish) [reply infoschema/tables_test.go:620]: Sounds like we can match `xxxx:` to distinguish fields.

IMO, we may also need some regression tests to verify that the slow log can be correctly parsed by percona tools.

File: util/execdetails/execdetails.go (hunk 1 above)
Comment (crazycs520) [reply util/execdetails/execdetails.go:152]: ping~ @bb7133 @qw4990 @breeswish PTAL

File: infoschema/tables_test.go (hunk 2 above)
Comment (crazycs520) [reply infoschema/tables_test.go:620]: match `xxxx:` to distinguish fields is a great idea. I like it.

Comment (sylzd) [reply infoschema/tables_test.go:620]: > Sounds like we can match `xxxx:` to distinguish fields.
> 
> IMO, we may also need some regression tests to verify that the slow log can be correctly parsed by percona tools.

LGTM, I will add this pattern.

File: util/execdetails/execdetails.go (hunk 1 above)
Comment (breeswish) [reply util/execdetails/execdetails.go:152]: This collapse looks great!

**Comment by breeswish** [comment]
Idea: As our [doc](https://docs.pingcap.com/tidb/stable/identify-slow-queries#parse-tidb-slow-logs-with-pt-query-digest) suggests to use pt-query-digest to parse the slow query, how about we ensure compatibility with it?

@crazycs520

**Comment by crazycs520** [comment]
> Idea: As our [doc](https://docs.pingcap.com/tidb/stable/identify-slow-queries#parse-tidb-slow-logs-with-pt-query-digest) suggests to use pt-query-digest to parse the slow query, how about we _additionally_ ensure compatibility with it?
> 
> @crazycs520

Great, I will think about how to add a test about use `pt-query-digest` to parse our slow log.

File: util/execdetails/execdetails.go (hunk 1 above)
Comment (sylzd) [reply util/execdetails/execdetails.go:152]: Does the order matters?

Comment (crazycs520) [reply util/execdetails/execdetails.go:152]: Maybe order by the count with desc is easier to read.

File: infoschema/tables_test.go (hunk 2 above)
Comment (sylzd) [reply infoschema/tables_test.go:620]: I made another solution about this , but I am not sure if it is better since it used `regex`.

File: util/execdetails/execdetails.go (hunk 1 above)
Comment (crazycs520) [reply util/execdetails/execdetails.go:152]: @sylzd Please resolve this comment~

Comment (sylzd) [reply util/execdetails/execdetails.go:152]: It does look Great, but what i worried is that the compatibility maybe broken more, which zoom in the radius of damage.
like this line:
https://github.com/pingcap/tidb/blob/e5bf177cd91ad689176c5f647e1e37d5cd47899c/util/stmtsummary/statement_summary.go#L763

File: infoschema/tables_test.go (hunk 2 above)
Comment (sylzd) [reply infoschema/tables_test.go:620]: What if we make both parser exits, when ',' exits, we used the new parser. To old version, we use the regex parser with little perf loss.🤔

File: util/execdetails/execdetails.go (hunk 1 above)
Comment (crazycs520) [reply util/execdetails/execdetails.go:152]: why this change will affect this?

Comment (sylzd) [reply util/execdetails/execdetails.go:152]: Aha~ My mistake, `BackoffTypes` here seems got from stats in memory instead of slow.log.  this case is not suitable, what I mean is that some tools like `pt-query-digest` or some application on table `slow_query` may be a little confused~  

BTW, I am not very sure, what's the purpose of the backoffTypes shown in slow log. 🤣

File: infoschema/tables_test.go (hunk 2 above)
Comment (sylzd) [reply infoschema/tables_test.go:620]: I think about the solution once again. I prefer the `comma added` solution, reasons are like that:
1. clear format makes it easy to parse by other tools.
2. compatibility problem above is a small probability event, and even it happens, we can easily solve it by `flush logs`.
3. slow log can rotate by the default config, which makes the little bug disappear when time goes on.

**Comment by sylzd** [comment]
#28474 is a better way.
